File: backend/services/data-cleaning-service/src/main/java/com/datamate/cleaning/application/CleaningTaskService.java
package com.datamate.cleaning.application;


import com.datamate.cleaning.application.scheduler.CleaningTaskScheduler;
import com.datamate.cleaning.common.enums.CleaningTaskStatusEnum;
import com.datamate.cleaning.common.enums.ExecutorType;
import com.datamate.cleaning.domain.model.TaskProcess;
import com.datamate.cleaning.domain.repository.CleaningResultRepository;
import com.datamate.cleaning.domain.repository.CleaningTaskRepository;
import com.datamate.cleaning.domain.repository.OperatorInstanceRepository;
import com.datamate.cleaning.infrastructure.validator.CleanTaskValidator;
import com.datamate.cleaning.interfaces.dto.*;
import com.datamate.common.infrastructure.exception.BusinessException;
import com.datamate.common.infrastructure.exception.SystemErrorCode;
import com.datamate.common.interfaces.PagedResponse;
import com.datamate.common.interfaces.PagingQuery;
import com.datamate.datamanagement.application.DatasetApplicationService;
import com.datamate.datamanagement.application.DatasetFileApplicationService;
import com.datamate.datamanagement.common.enums.DatasetType;
import com.datamate.datamanagement.domain.model.dataset.Dataset;
import com.datamate.datamanagement.domain.model.dataset.DatasetFile;
import com.datamate.datamanagement.interfaces.dto.CreateDatasetRequest;
import com.datamate.operator.domain.repository.OperatorRepository;
import com.datamate.operator.infrastructure.exception.OperatorErrorCode;
import com.datamate.operator.interfaces.dto.OperatorDto;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
@Service
@RequiredArgsConstructor
public class CleaningTaskService {
    private final CleaningTaskRepository cleaningTaskRepo;

    private final OperatorInstanceRepository operatorInstanceRepo;

    private final OperatorRepository operatorRepo;

    private final CleaningResultRepository cleaningResultRepo;

    private final CleaningTaskScheduler taskScheduler;

    private final DatasetApplicationService datasetService;

    private final DatasetFileApplicationService datasetFileService;

    private final CleanTaskValidator cleanTaskValidator;

    private final String DATASET_PATH = "/dataset";

    private final String FLOW_PATH = "/flow";

    private static final Pattern STANDARD_LEVEL_PATTERN = Pattern.compile(
            "\\b(DEBUG|Debug|INFO|Info|WARN|Warn|WARNING|Warning|ERROR|Error|FATAL|Fatal)\\b"
    );

    private static final Pattern EXCEPTION_SUFFIX_PATTERN = Pattern.compile(
            "\\b\\w+(Warning|Error|Exception)\\b"
    );

    private final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    public List<CleaningTaskDto> getTasks(String status, String keywords, Integer page, Integer size) {
        List<CleaningTaskDto> tasks = cleaningTaskRepo.findTasks(status, keywords, page, size);
        tasks.forEach(this::setProcess);
        return tasks;
    }

    private void setProcess(CleaningTaskDto task) {
        int[] count = cleaningResultRepo.countByInstanceId(task.getId());
        task.setProgress(CleaningProcess.of(task.getFileCount(), count[0], count[1]));
    }

    public int countTasks(String status, String keywords) {
        return cleaningTaskRepo.findTasks(status, keywords, null, null).size();
    }

    @Transactional
    public CleaningTaskDto createTask(CreateCleaningTaskRequest request) {
        cleanTaskValidator.checkNameDuplication(request.getName());
        cleanTaskValidator.checkInputAndOutput(request.getInstance());

        ExecutorType executorType = cleanTaskValidator.checkAndGetExecutorType(request.getInstance());

        Dataset destDataset;
        if (StringUtils.isNotBlank(request.getDestDatasetId())) {
            destDataset = datasetService.getDataset(request.getDestDatasetId());
        } else {
            CreateDatasetRequest createDatasetRequest = new CreateDatasetRequest();
            createDatasetRequest.setName(request.getDestDatasetName());
            createDatasetRequest.setDatasetType(DatasetType.valueOf(request.getDestDatasetType()));
            createDatasetRequest.setStatus("ACTIVE");
            destDataset = datasetService.createDataset(createDatasetRequest);
        }
        Dataset srcDataset = datasetService.getDataset(request.getSrcDatasetId());

        CleaningTaskDto task = new CleaningTaskDto();
        task.setName(request.getName());
        task.setDescription(request.getDescription());
        task.setStatus(CleaningTaskStatusEnum.PENDING);
        String taskId = UUID.randomUUID().toString();
        task.setId(taskId);
        task.setSrcDatasetId(request.getSrcDatasetId());
        task.setSrcDatasetName(request.getSrcDatasetName());
        task.setDestDatasetId(destDataset.getId());
        task.setDestDatasetName(destDataset.getName());
        task.setBeforeSize(srcDataset.getSizeBytes());
        task.setFileCount(srcDataset.getFileCount().intValue());
        cleaningTaskRepo.insertTask(task);

        operatorInstanceRepo.insertInstance(taskId, request.getInstance());

        prepareTask(task, request.getInstance(), executorType);
        scanDataset(taskId, request.getSrcDatasetId());
        taskScheduler.executeTask(taskId);
        return task;
    }

    public CleaningTaskDto getTask(String taskId) {
        CleaningTaskDto task = cleaningTaskRepo.findTaskById(taskId);
        setProcess(task);
        task.setInstance(operatorInstanceRepo.findOperatorByInstanceId(taskId));
        return task;
    }

    public List<CleaningResultDto> getTaskResults(String taskId) {
        return cleaningResultRepo.findByInstanceId(taskId);
    }

    public List<CleaningTaskLog> getTaskLog(String taskId) {
        cleanTaskValidator.checkTaskId(taskId);
        String logPath = FLOW_PATH + "/" + taskId + "/output.log";
        try (Stream<String> lines = Files.lines(Paths.get(logPath))) {
            List<CleaningTaskLog> logs = new ArrayList<>();
            AtomicReference<String> lastLevel = new AtomicReference<>("INFO");
            lines.forEach(line -> {
                lastLevel.set(getLogLevel(line, lastLevel.get()));
                CleaningTaskLog log = new CleaningTaskLog();
                log.setLevel(lastLevel.get());
                log.setMessage(line);
                logs.add(log);
            });
            return logs;
        } catch (IOException e) {
            log.error("Fail to read log file {}", logPath, e);
            return Collections.emptyList();
        }
    }

    private String getLogLevel(String logLine, String defaultLevel) {
        if (logLine == null || logLine.trim().isEmpty()) {
            return defaultLevel;
        }

        Matcher stdMatcher = STANDARD_LEVEL_PATTERN.matcher(logLine);
        if (stdMatcher.find()) {
            return stdMatcher.group(1).toUpperCase();
        }

        Matcher exMatcher = EXCEPTION_SUFFIX_PATTERN.matcher(logLine);
        if (exMatcher.find()) {
            String match = exMatcher.group(1).toUpperCase();
            if ("WARNING".equals(match)) return "WARN";
            if ("ERROR".equals(match) || "EXCEPTION".equals(match)) return "ERROR";
        }
        return defaultLevel;
    }

    @Transactional
    public void deleteTask(String taskId) {
        cleanTaskValidator.checkTaskId(taskId);
        cleaningTaskRepo.deleteTaskById(taskId);
        operatorInstanceRepo.deleteByInstanceId(taskId);
        cleaningResultRepo.deleteByInstanceId(taskId);
        try {
            FileUtils.deleteDirectory(new File(FLOW_PATH + "/" + taskId));
        } catch (IOException e) {
            log.warn("Can't delete flow path with task id: {}.", taskId, e);
        }
    }

    public void executeTask(String taskId) {
        List<CleaningResultDto> succeed = cleaningResultRepo.findByInstanceId(taskId, "COMPLETED");
        Set<String> succeedSet = succeed.stream().map(CleaningResultDto::getSrcFileId).collect(Collectors.toSet());
        CleaningTaskDto task = cleaningTaskRepo.findTaskById(taskId);
        scanDataset(taskId, task.getSrcDatasetId(), succeedSet);
        cleaningResultRepo.deleteByInstanceId(taskId, "FAILED");
        taskScheduler.executeTask(taskId);
    }

    private void prepareTask(CleaningTaskDto task, List<OperatorInstanceDto> instances, ExecutorType executorType) {
        List<OperatorDto> allOperators = operatorRepo.findAllOperators();
        Map<String, OperatorDto> operatorDtoMap = allOperators.stream()
                .collect(Collectors.toMap(OperatorDto::getId, Function.identity()));

        TaskProcess process = new TaskProcess();
        process.setInstanceId(task.getId());
        process.setDatasetId(task.getDestDatasetId());
        process.setExecutorType(executorType.getValue());
        process.setDatasetPath(FLOW_PATH + "/" + task.getId() + "/dataset.jsonl");
        process.setExportPath(DATASET_PATH + "/" + task.getDestDatasetId());
        process.setProcess(instances.stream()
                .map(instance -> {
                    OperatorDto operatorDto = operatorDtoMap.get(instance.getId());
                    Map<String, Object> stringObjectMap = getDefaultValue(operatorDto);
                    stringObjectMap.putAll(instance.getOverrides());
                    Map<String, Object> runtime = getRuntime(operatorDto);
                    stringObjectMap.putAll(runtime);
                    return Map.of(instance.getId(), stringObjectMap);
                })
                .toList());

        ObjectMapper jsonMapper = new ObjectMapper(new YAMLFactory());
        jsonMapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
        JsonNode jsonNode = jsonMapper.valueToTree(process);

        DumperOptions options = new DumperOptions();
        options.setIndent(2);
        options.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
        Yaml yaml = new Yaml(options);

        File file = new File(FLOW_PATH + "/" + task.getId() + "/process.yaml");
        file.getParentFile().mkdirs();

        try (FileWriter writer = new FileWriter(file)) {
            yaml.dump(jsonMapper.treeToValue(jsonNode, Map.class), writer);
        } catch (IOException e) {
            log.error("Failed to prepare process.yaml.", e);
            throw BusinessException.of(SystemErrorCode.FILE_SYSTEM_ERROR);
        }
    }

    private Map<String, Object> getDefaultValue(OperatorDto operatorDto) {
        if (StringUtils.isBlank(operatorDto.getSettings())) {
            return new HashMap<>();
        }

        Map<String, Object> defaultSettings = new HashMap<>();
        try {
            Map<String, Map<String, Object>> settings = OBJECT_MAPPER.readValue(operatorDto.getSettings(), Map.class);
            for  (Map.Entry<String, Map<String, Object>> entry : settings.entrySet()) {
                String key = entry.getKey();
                Map<String, Object> setting = entry.getValue();
                String type = setting.get("type").toString();
                switch (type) {
                    case "slider":
                    case "switch":
                    case "select":
                    case "input":
                    case "radio":
                    case "checkbox":
                        if (setting.containsKey("defaultVal")) {
                            defaultSettings.put(key, setting.get("defaultVal"));
                        }
                        break;
                    case "range":
                        List<Object> rangeDefault = getRangeDefault(setting);
                        if (CollectionUtils.isNotEmpty(rangeDefault)) {
                            defaultSettings.put(key, rangeDefault);
                        }
                        break;
                    default:
                }
            }
            return defaultSettings;
        } catch (JsonProcessingException e) {
            throw BusinessException.of(OperatorErrorCode.SETTINGS_PARSE_FAILED, e.getMessage());
        }
    }

    private List<Object> getRangeDefault(Map<String, Object> setting) {
        List<Object> defaultValue = new ArrayList<>();
        Object properties = setting.get("properties");
        if (properties instanceof List<?> list) {
            for (Object o : list) {
                Map<String, Object> map = OBJECT_MAPPER.convertValue(o, Map.class);
                if (map.containsKey("defaultVal")) {
                    defaultValue.add(map.get("defaultVal"));
                }
            }
        }
        return defaultValue;
    }

    private Map<String, Object> getRuntime(OperatorDto operatorDto) {
        if (StringUtils.isBlank(operatorDto.getRuntime())) {
            return new HashMap<>();
        }
        try {
            return OBJECT_MAPPER.readValue(operatorDto.getRuntime(), Map.class);
        } catch (JsonProcessingException e) {
            throw BusinessException.of(OperatorErrorCode.SETTINGS_PARSE_FAILED, e.getMessage());
        }
    }

    private void scanDataset(String taskId, String srcDatasetId) {
        doScan(taskId, srcDatasetId, file -> true);
    }

    private void scanDataset(String taskId, String srcDatasetId, Set<String> succeedFiles) {
        doScan(taskId, srcDatasetId, file -> !succeedFiles.contains(file.getId()));
    }

    private void doScan(String taskId, String srcDatasetId, Predicate<DatasetFile> filterCondition) {
        cleanTaskValidator.checkTaskId(taskId);
        String targetFilePath = FLOW_PATH + "/" + taskId + "/dataset.jsonl";
        File targetFile = new File(targetFilePath);
        if (targetFile.getParentFile() != null && !targetFile.getParentFile().exists()) {
            targetFile.getParentFile().mkdirs();
        }

        int pageNumber = 0;
        int pageSize = 500;
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(targetFile))) {
            PagedResponse<DatasetFile> datasetFiles;
            do {
                PagingQuery pageRequest = new PagingQuery(pageNumber, pageSize);
                datasetFiles = datasetFileService.getDatasetFiles(srcDatasetId, null, null, null, pageRequest);
                if (datasetFiles.getContent().isEmpty()) {
                    break;
                }
                for (DatasetFile content : datasetFiles.getContent()) {
                    if (!filterCondition.test(content)) {
                        continue;
                    }
                    Map<String, Object> fileMap = Map.of(
                            "fileName", content.getFileName(),
                            "fileSize", content.getFileSize(),
                            "filePath", content.getFilePath(),
                            "fileType", content.getFileType(),
                            "fileId", content.getId()
                    );
                    writer.write(OBJECT_MAPPER.writeValueAsString(fileMap));
                    writer.newLine();
                }
                pageNumber++;
            } while (pageNumber < datasetFiles.getTotalPages());
        } catch (IOException e) {
            log.error("Failed to write dataset.jsonl for taskId: {}", taskId, e);
            throw BusinessException.of(SystemErrorCode.FILE_SYSTEM_ERROR);
        }
    }

    public void stopTask(String taskId) {
        taskScheduler.stopTask(taskId);
    }

    public List<OperatorInstanceDto> getInstanceByTemplateId(String templateId) {
        return operatorInstanceRepo.findInstanceByInstanceId(templateId);
    }
}

File: runtime/datamate-python/app/module/cleaning/service/cleaning_task_service.py
import json
import re
import shutil
import uuid
from pathlib import Path
from typing import List, Dict, Any, Set

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.logging import get_logger
from app.db.models.base_entity import LineageNode, LineageEdge
from app.core.exception import BusinessError, ErrorCodes
from app.module.cleaning.repository import (
    CleaningTaskRepository,
    CleaningResultRepository,
    OperatorInstanceRepository,
)
from app.module.cleaning.schema import (
    CleaningTaskDto,
    CreateCleaningTaskRequest,
    CleaningResultDto,
    CleaningTaskLog,
    OperatorInstanceDto,
    CleaningProcess,
    CleaningTaskStatus,
)
from app.module.cleaning.service.cleaning_task_validator import CleaningTaskValidator
from app.module.cleaning.service.cleaning_task_scheduler import CleaningTaskScheduler
from app.module.shared.common.lineage import LineageService
from app.module.shared.schema.lineage import NodeType, EdgeType

logger = get_logger(__name__)

DATASET_PATH = "/dataset"
FLOW_PATH = "/flow"


class CleaningTaskService:
    """Service for managing cleaning tasks"""

    def __init__(
        self,
        task_repo: CleaningTaskRepository,
        result_repo: CleaningResultRepository,
        operator_instance_repo: OperatorInstanceRepository,
        operator_service,
        scheduler: CleaningTaskScheduler,
        validator: CleaningTaskValidator,
        dataset_service,
        lineage_service: LineageService,
    ):
        self.task_repo = task_repo
        self.result_repo = result_repo
        self.operator_instance_repo = operator_instance_repo
        self.operator_service = operator_service
        self.scheduler = scheduler
        self.validator = validator
        self.dataset_service = dataset_service
        self.lineage_service = lineage_service

    async def get_tasks(
        self,
        db: AsyncSession,
        status: str | None = None,
        keyword: str | None = None,
        page: int | None = None,
        size: int | None = None,
    ) -> List[CleaningTaskDto]:
        """Get cleaning tasks"""
        tasks = await self.task_repo.find_tasks(db, status, keyword, page, size)

        for task in tasks:
            await self._set_process(db, task)

        return tasks

    async def _set_process(self, db: AsyncSession, task: CleaningTaskDto) -> None:
        """Set task progress"""
        completed, failed = await self.result_repo.count_by_instance_id(db, task.id)
        task.progress = CleaningProcess.of(task.file_count or 0, completed, failed)

    async def count_tasks(
        self,
        db: AsyncSession,
        status: str | None = None,
        keyword: str | None = None,
    ) -> int:
        """Count cleaning tasks"""
        tasks = await self.task_repo.find_tasks(db, status, keyword, None, None)
        return len(tasks)

    async def get_task(self, db: AsyncSession, task_id: str) -> CleaningTaskDto:
        """Get task by ID"""
        task = await self.task_repo.find_task_by_id(db, task_id)
        if not task:
            raise BusinessError(ErrorCodes.CLEANING_TASK_NOT_FOUND, task_id)

        await self._set_process(db, task)

        instances = await self.operator_instance_repo.find_operator_by_instance_id(db, task_id)

        # Batch query operators
        all_operators = await self.operator_service.get_operators(db=db, page=0, size=1000, categories=[], keyword=None,
                                                                  is_star=None)
        operator_map = {op.id: op for op in all_operators}

        task.instance = []
        for inst in instances:
            operator = operator_map.get(inst.operator_id)
            if operator:
                task.instance.append(OperatorInstanceDto(
                    id=operator.id,
                    name=operator.name,
                    description=operator.description,
                    inputs=operator.inputs,
                    outputs=operator.outputs,
                    settings=operator.settings,
                    categories=operator.categories,
                ))
            else:
                task.instance.append(OperatorInstanceDto(id=inst.operator_id))

        return task

    async def create_task(
        self,
        db: AsyncSession,
        request: CreateCleaningTaskRequest
    ) -> CleaningTaskDto:
        """Create new cleaning task"""
        if request.instance and request.template_id:
            instances = await self.get_instance_by_template_id(db, request.template_id)
            request.instance = instances

        await self.validator.check_task_name_duplication(db, request.name)
        self.validator.check_input_and_output(request.instance)
        executor_type = self.validator.check_and_get_executor_type(request.instance)

        task_id = str(uuid.uuid4())

        dest_dataset_id = request.dest_dataset_id
        dest_dataset_name = request.dest_dataset_name

        if not dest_dataset_id:
            logger.info(f"Creating new dataset: {dest_dataset_name}, type: {request.dest_dataset_type}")
            dest_dataset_response = await self.dataset_service.create_dataset(
                name=dest_dataset_name,
                dataset_type=request.dest_dataset_type,
                description="",
                status="ACTIVE"
            )
            dest_dataset_id = dest_dataset_response.id
            logger.info(f"Successfully created dataset: {dest_dataset_id}")
        else:
            logger.info(f"Using existing dataset: {dest_dataset_id}")
            dest_dataset_response = await self.dataset_service.get_dataset(dest_dataset_id)

        src_dataset = await self.dataset_service.get_dataset(request.src_dataset_id)
        if not src_dataset:
            raise BusinessError(ErrorCodes.CLEANING_DATASET_NOT_FOUND, request.src_dataset_id)

        task_dto = CleaningTaskDto(
            id=task_id,
            name=request.name,
            description=request.description,
            status=CleaningTaskStatus.PENDING,
            src_dataset_id=request.src_dataset_id,
            src_dataset_name=request.src_dataset_name,
            dest_dataset_id=dest_dataset_id,
            dest_dataset_name=dest_dataset_name,
            before_size=src_dataset.totalSize,
            file_count=src_dataset.fileCount,
            retry_count=-1,
        )

        await self.task_repo.insert_task(db, task_dto)

        await self._add_cleaning_to_graph(src_dataset, task_dto, dest_dataset_response)

        await self.operator_instance_repo.insert_instance(db, task_id, request.instance)

        all_operators = await self.operator_service.get_operators(db=db, page=0, size=1000, categories=[], keyword=None, is_star=None)
        operator_map = {op.id: op for op in all_operators}

        await self.prepare_task(dest_dataset_id, task_id, request.instance, operator_map, executor_type)

        return await self.get_task(db, task_id)

    async def _add_cleaning_to_graph(
        self,
        src_dataset,
        task: CleaningTaskDto,
        dest_dataset,
    ) -> None:
        """
        添加清洗任务到血缘图
        """
        from_node = LineageNode(
            id=src_dataset.id,
            node_type=NodeType.DATASET.value,
            name=src_dataset.name,
            description=src_dataset.description or "",
        )

        to_node = LineageNode(
            id=dest_dataset.id,
            node_type=NodeType.DATASET.value,
            name=dest_dataset.name,
            description=dest_dataset.description or "",
        )

        edge = LineageEdge(
            process_id=task.id,
            name=task.name or "",
            description=task.description or "",
            edge_type=EdgeType.DATA_CLEANING.value,
            from_node_id=from_node.id,
            to_node_id=to_node.id,
        )

        await self.lineage_service.generate_graph(from_node, edge, to_node)

    async def prepare_task(
        self,
        dataset_id: str,
        task_id: str,
        instances: List[OperatorInstanceDto],
        operator_map: dict,
        executor_type: str,
    ) -> None:
        """Prepare task configuration file"""
        process_config = {
            "dataset_id": dataset_id,
            "instance_id": task_id,
            "dataset_path": f"{FLOW_PATH}/{task_id}/dataset.jsonl",
            "export_path": f"{DATASET_PATH}/{dataset_id}",
            "executor_type": executor_type,
            "process": [],
        }

        for instance in instances:
            operator = operator_map.get(instance.id)
            if not operator:
                continue

            operator_config = self._get_default_values(operator)
            operator_config.update(instance.overrides)

            runtime_config = self._get_runtime_config(operator)
            operator_config.update(runtime_config)

            process_config["process"].append({instance.id: operator_config})

        config_file_path = Path(f"{FLOW_PATH}/{task_id}/process.yaml")
        config_file_path.parent.mkdir(parents=True, exist_ok=True)

        import yaml
        try:
            with open(config_file_path, 'w', encoding='utf-8') as f:
                yaml.dump(process_config, f, default_flow_style=False, allow_unicode=True)
        except Exception as e:
            logger.error(f"Failed to write process.yaml: {e}")
            raise BusinessError(ErrorCodes.CLEANING_FILE_SYSTEM_ERROR, str(e))

    def _get_default_values(self, operator) -> Dict[str, Any]:
        """Get default values from operator settings"""
        if not operator.settings:
            return {}

        try:
            settings = json.loads(operator.settings)
            defaults = {}

            for key, value in settings.items():
                setting_type = value.get("type")
                if "defaultVal" in value:
                    defaults[key] = value["defaultVal"]

            return defaults
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse settings: {e}")
            return {}

    def _get_runtime_config(self, operator) -> Dict[str, Any]:
        """Get runtime configuration from operator"""
        if not operator.runtime:
            return {}

        try:
            return json.loads(operator.runtime)
        except json.JSONDecodeError as e:
            logger.error(f"Failed to parse runtime config: {e}")
            return {}

    async def scan_dataset(
        self,
        db: AsyncSession,
        task_id: str,
        src_dataset_id: str,
        succeed_files: Set[str] | None = None,
    ) -> None:
        """Scan source dataset and create dataset.jsonl"""
        target_file_path = Path(f"{FLOW_PATH}/{task_id}/dataset.jsonl")
        target_file_path.parent.mkdir(parents=True, exist_ok=True)

        query = text("""
            SELECT id, file_name, file_path, file_type, file_size
            FROM t_dm_dataset_files
            WHERE dataset_id = :dataset_id
            ORDER BY created_at
        """)

        result = await db.execute(query, {"dataset_id": src_dataset_id})
        files = result.fetchall()

        with open(target_file_path, 'w', encoding='utf-8') as f:
            for file in files:
                if succeed_files and file.id in succeed_files:
                    continue

                file_info = {
                    "fileId": file.id,
                    "fileName": file.file_name,
                    "filePath": file.file_path,
                    "fileType": file.file_type,
                    "fileSize": file.file_size,
                }
                f.write(json.dumps(file_info, ensure_ascii=False) + "\n")

    async def get_task_results(self, db: AsyncSession, task_id: str) -> List[CleaningResultDto]:
        """Get task results"""
        return await self.result_repo.find_by_instance_id(db, task_id)

    async def get_task_log(self, db: AsyncSession, task_id: str, retry_count: int) -> List[CleaningTaskLog]:
        """Get task log"""
        self.validator.check_task_id(task_id)

        log_path = Path(f"{FLOW_PATH}/{task_id}/output.log")
        if retry_count > 0:
            log_path = Path(f"{FLOW_PATH}/{task_id}/output.log.{retry_count}")

        if not log_path.exists():
            return []

        logs = []
        last_level = "INFO"

        standard_level_pattern = re.compile(
            r"\b(DEBUG|Debug|INFO|Info|WARN|Warn|WARNING|Warning|ERROR|Error|FATAL|Fatal)\b"
        )
        exception_suffix_pattern = re.compile(r"\b\w+(Warning|Error|Exception)\b")

        with open(log_path, 'r', encoding='utf-8') as f:
            for line in f:
                last_level = self._get_log_level(line, last_level, standard_level_pattern, exception_suffix_pattern)
                logs.append(CleaningTaskLog(level=last_level, message=line.rstrip()))

        return logs

    def _get_log_level(self, line: str, default_level: str, std_pattern, ex_pattern) -> str:
        """Extract log level from log line"""
        if not line or not line.strip():
            return default_level

        std_match = std_pattern.search(line)
        if std_match:
            return std_match.group(1).upper()

        ex_match = ex_pattern.search(line)
        if ex_match:
            match = ex_match.group(1).upper()
            if match == "WARNING":
                return "WARN"
            if match in ["ERROR", "EXCEPTION"]:
                return "ERROR"

        return default_level

    async def delete_task(self, db: AsyncSession, task_id: str) -> None:
        """Delete task"""
        self.validator.check_task_id(task_id)

        await self.task_repo.delete_task_by_id(db, task_id)
        await self.operator_instance_repo.delete_by_instance_id(db, task_id)
        await self.result_repo.delete_by_instance_id(db, task_id)

        task_path = Path(f"{FLOW_PATH}/{task_id}")
        if task_path.exists():
            try:
                shutil.rmtree(task_path)
            except Exception as e:
                logger.warning(f"Failed to delete task path {task_id}: {e}")

    async def execute_task(self, db: AsyncSession, task_id: str) -> bool:
        """Execute task"""
        succeeded = await self.result_repo.find_by_instance_id(db, task_id, "COMPLETED")
        succeed_set = {res.src_file_id for res in succeeded}

        task = await self.task_repo.find_task_by_id(db, task_id)
        if not task:
            raise BusinessError(ErrorCodes.CLEANING_TASK_NOT_FOUND, task_id)

        await self.scan_dataset(db, task_id, task.src_dataset_id, succeed_set)
        await self.result_repo.delete_by_instance_id(db, task_id, "FAILED")

        return await self.scheduler.execute_task(db, task_id, (task.retry_count or 0) + 1)

    async def stop_task(self, db: AsyncSession, task_id: str) -> bool:
        """Stop task"""
        return await self.scheduler.stop_task(db, task_id)

    async def get_instance_by_template_id(
        self,
        db: AsyncSession,
        template_id: str
    ) -> List[OperatorInstanceDto]:
        """Get instances by template ID (delegated to template service)"""
        instances = await self.operator_instance_repo.find_operator_by_instance_id(db, template_id)

        # Batch query operators
        all_operators = await self.operator_service.get_operators(db=db, page=0, size=1000, categories=[], keyword=None,
                                                                  is_star=None)
        operator_map = {op.id: op for op in all_operators}

        result = []
        for inst in instances:
            operator = operator_map.get(inst.operator_id)
            if operator:
                operator_dto = OperatorInstanceDto(
                    id=operator.id,
                    name=operator.name,
                    description=operator.description,
                    inputs=operator.inputs,
                    outputs=operator.outputs,
                    settings=operator.settings,
                    categories=operator.categories,
                )
                if inst.settings_override:
                    try:
                        operator_dto.overrides = json.loads(inst.settings_override)
                    except json.JSONDecodeError as e:
                        logger.error(f"Failed to parse settings for {inst.operator_id}: {e}")
                result.append(operator_dto)

        return result
